### DISCUSSION TITLE
Message Order Not Preserved

### DISCUSSION THREAD

**Discussion by Donadominic** [nanomq/nanomq#896] [Q&A] [answered]
While using NanoMQ when the client is disconnected and connected back again the order in which the messages are published is not preserved. It is reading data in some random order. But when the client is connected the messages are received in proper order. How to receive messages in proper order?

I was sending a sequence of messages when the client was disconnected. And when the client is connected back again I am expecting the messages to be received in the proper order they are send.

Publisher Code
public class Publisher { public static void main(String[] args) {

    try {
        //publishing a message
        MqttClient client=new MqttClient("tcp://192.168.2.8:1885",MqttClient.generateClientId(),new MemoryPersistence());
        MqttConnectOptions options = new MqttConnectOptions();
        options.setCleanSession(false);
        options.setUserName("user1");
        options.setPassword("password".toCharArray());
        client.connect(options);
        
        MqttMessage message=new MqttMessage();
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");

        for(int i=1;i<=20;i++) {
            Date date = new Date();
            String str=simpleDateFormat.format(date)+"    message "+i;
            message.setPayload(str.getBytes());
            message.setQos(2);
            message.setRetained(true);
            client.publish("iot_data_1",message);
        }
        
        client.disconnect();

    } catch(MqttException me) {
        System.out.println(me);
    }

}
}

Consumer Code
public class Subscriber3 { public static void main(String[]args) throws MqttException { MqttClient client =new MqttClient("tcp://192.168.2.8:1885","subscriber004",new MemoryPersistence()); MqttConnectOptions options=new MqttConnectOptions(); options.setCleanSession(false); options.setUserName("user009"); options.setPassword("password008".toCharArray()); client.connect(options);

    client.setCallback(new MqttCallback() {
        @Override
        public void connectionLost(Throwable cause) {
            System.out.println("Connection lost");
        }

        @Override
        public void messageArrived(String topic, MqttMessage message) throws Exception {
            System.out.println(new String(message.getPayload()));
        }

        @Override
        public void deliveryComplete(IMqttDeliveryToken token) {
            System.out.println("Delivery Complete");
        }
    });

    client.subscribe("iot_data_1");
    //client.disconnect();
}
}

But messages are received in some random order. All the messages are received. But I need them in the order they are published.

NanoMQ latest version
Java Client in Windows and NanoMq on ubuntu centos in docker container
Java- Eclipse Paho
Publishing a sequence of messages when the client is disconnected and when client is connected back need persisted messages in order

I need messages in the order in which they are published not in some random order.

**Comment by JaylinYu** [comment]
Hi, Thanks for reporting, this belongs to ISSUE page.
And This is a known issue, I am planning to fix this in next version.

**Reply by Donadominic** [reply]
When can we expect the release?

**Reply by JaylinYu** [reply]
should be out next month

**Reply by Donadominic** [reply]
We are planning to start building a system in January. We are looking forward towards the new release of nanomq with message order preservation. Can you please provide an exact date of release so that we can start working on the project.

**Reply by JaylinYu** [reply]
0.14 gonna be released before Jan for sure, But I don't quite sure which day of Dec, Thank you for your demand. Any other features is required from nanomq in your next system build?

**Comment by JaylinYu** [comment, marked as answer]
Plz check ver 0.13.3

**Comment by JaylinYu** [comment]
@Donadominic plz mark the answer if this fixes your issue

**Reply by Donadominic** [reply]
More than processing messages in order. We also want to consume the already published messages in batches instead of pushing them one by one to the client side. But with mqtt batch consuming is not possible.

**Reply by JaylinYu** [reply]
understood, It might be possible to implement a feature in rule engine that merge multiple msg' payload into one, is it being helpful?
BTW, is the fix work for you?

**Reply by Donadominic** [reply]
Yes. Now messages are received in order.

**Reply by JaylinYu** [reply]
thanks
Plz mark the answer.

**Reply by JaylinYu** [reply]
Hi @Donadominic 
I am planning to support batch resend, plz fire a feature request issue with detailed explanation
